File: Mockingjay/NSURLSessionConfiguration.m
//
//  NSURLSessionConfiguration+Mockingjay.m
//  Mockingjay
//
//  Created by Kyle Fuller on 01/03/2015.
//  Copyright (c) 2015 Cocode. All rights reserved.
//

#import <objc/objc-runtime.h>
#import <Foundation/Foundation.h>
#import <Mockingjay/Mockingjay-Swift.h>

// TODO Rewrite in Swift.

@interface NSURLSessionConfiguration (Mockingjay)

@end

@implementation NSURLSessionConfiguration (Mockingjay)

+ (void)initialize {
  if (self == [NSURLSessionConfiguration class]) {
    Method defaultSessionConfiguration = class_getClassMethod(self, @selector(defaultSessionConfiguration));
    Method mockingjayDefaultSessionConfiguration = class_getClassMethod(self, @selector(mockingjayDefaultSessionConfiguration));
    method_exchangeImplementations(defaultSessionConfiguration, mockingjayDefaultSessionConfiguration);

    Method ephemeralSessionConfiguration = class_getClassMethod(self, @selector(ephemeralSessionConfiguration));
    Method mockingjayEphemeralSessionConfiguration = class_getClassMethod(self, @selector(mockingjayEphemeralSessionConfiguration));
    method_exchangeImplementations(ephemeralSessionConfiguration, mockingjayEphemeralSessionConfiguration);
  }
}

+ (NSURLSessionConfiguration *)mockingjayDefaultSessionConfiguration {
  NSURLSessionConfiguration *sessionConfiguration = [self mockingjayDefaultSessionConfiguration];
  NSMutableArray *protocolClasses = [sessionConfiguration.protocolClasses mutableCopy];
  [protocolClasses insertObject:MockingjayProtocol.class atIndex:0];
  sessionConfiguration.protocolClasses = protocolClasses;
  return sessionConfiguration;
}

+ (NSURLSessionConfiguration *)mockingjayEphemeralSessionConfiguration {
  NSURLSessionConfiguration *sessionConfiguration = [self mockingjayEphemeralSessionConfiguration];
  NSMutableArray *protocolClasses = [sessionConfiguration.protocolClasses mutableCopy];
  [protocolClasses insertObject:MockingjayProtocol.class atIndex:0];
  sessionConfiguration.protocolClasses = protocolClasses;
  return sessionConfiguration;
}

@end

File: Mockingjay/NSURLSessionConfiguration.swift
//
//  NSURLSessionConfiguration.swift
//  Mockingjay
//
//  Created by Kyle Fuller on 01/03/2015.
//  Copyright (c) 2015 Cocode. All rights reserved.
//

import Foundation

let swizzleDefaultSessionConfiguration: Void = {
  guard
    let defaultSessionConfiguration = class_getClassMethod(URLSessionConfiguration.self, #selector(getter: URLSessionConfiguration.default)),
    let mockingjayDefaultSessionConfiguration = class_getClassMethod(URLSessionConfiguration.self, #selector(URLSessionConfiguration.mockingjayDefaultSessionConfiguration))
    else { return }
  method_exchangeImplementations(defaultSessionConfiguration, mockingjayDefaultSessionConfiguration)

  guard
    let ephemeralSessionConfiguration = class_getClassMethod(URLSessionConfiguration.self, #selector(getter: URLSessionConfiguration.ephemeral)),
    let mockingjayEphemeralSessionConfiguration = class_getClassMethod(URLSessionConfiguration.self, #selector(URLSessionConfiguration.mockingjayEphemeralSessionConfiguration))
    else { return }
    method_exchangeImplementations(ephemeralSessionConfiguration, mockingjayEphemeralSessionConfiguration)
}()

extension URLSessionConfiguration {
  /// Swizzles NSURLSessionConfiguration's default and ephermeral sessions to add Mockingjay
  @objc public class func mockingjaySwizzleDefaultSessionConfiguration() {
    _ = swizzleDefaultSessionConfiguration
  }

  @objc class func mockingjayDefaultSessionConfiguration() -> URLSessionConfiguration {
    let configuration = mockingjayDefaultSessionConfiguration()
    configuration.protocolClasses = [MockingjayProtocol.self] as [AnyClass] + configuration.protocolClasses!
    return configuration
  }

  @objc class func mockingjayEphemeralSessionConfiguration() -> URLSessionConfiguration {
    let configuration = mockingjayEphemeralSessionConfiguration()
    configuration.protocolClasses = [MockingjayProtocol.self] as [AnyClass] + configuration.protocolClasses!
    return configuration
  }
}
